[Doc] Add Keycloak login and config docs (#93)
nebari-dev/nebari-docs: docs/docs/how-tos/How to install and configure Keycloak.md

---
id: configuring-keycloak
title: How to configure Keycloak and add new users
description: |
  This is a how-to for configuring Keycloak and add new users.
---

## Configuring Keycloak

[Keycloak](https://www.keycloak.org/) is the name of the open-source user management software that's automatically deployed within Nebari. It's used to store the database of all

> users in your Nebari instance, and can provide connectivity to other services such as GitHub/Auth0 single sign-on.

Reviewer: I think here is a good place to mention the definitions of roles and groups in the context of Keycloak and what it provides. Then later, when they are mentioned, readers will have a general idea. Then at the end of the doc, we go into more details. Does that make sense?

> users in your Nebari instance, and can provide connectivity to other services such as GitHub/Auth0 single sign-on.

Author: uhm, not sure would like to keep it only for a summary of Keycloak. On the other hand, I think the `Adding a Nebari user` might be a good place to add a preview of that, something along the lines of:

> when you create your user using the Nebari Keycloak command, it will receive the basic permissions and roles associated with the `Analyst` group. For more information, check out the [Permissions and Roles] section below.

> users in your Nebari instance, and can provide connectivity to other services such as GitHub/Auth0 single sign-on.

Reviewer 2: What this doc needs is an intro section - like: what are folks learning? 

it goes from zero to configuring Keycloak, like in the other doc we need to add this at the top

> As with many user admin tools, in Keycloak you start with an initial login (`root`) that has the ability to administer and create new users. The `root` user is a Keycloak-specific user. It can only be used to login and manage the Keycloak identity management section of Nebari. It's not a user of the wider Nebari data science platform.

Author: we already mentioned this shortly in the deploy docs (almost certain about this...), so If possible, I would like this to be the source of truth for Keycloak.

> As with many user admin tools, in Keycloak you start with an initial login (`root`) that has the ability to administer and create new users. The `root` user is a Keycloak-specific user. It can only be used to login and manage the Keycloak identity management section of Nebari. It's not a user of the wider Nebari data science platform.

Reviewer 2: I would say both - since I assume folks will not linearly follow the docs, also redundancy. Someone updating keycloak configs might not be the same person deploying nebari

## Change Keycloak root password

> `root`'s password is generated by the `nebari init` command. If you ran this command while following the steps under [Installing Nebari](/getting-started/installing-nebari), you should have seen something like the following in your terminal output:

Author: I think I will re-write this sentence to make sure we don't mislead the user here. Thanks for pointing that out :)

```
Securely generated default random password=<alphanumeric-string> for Keycloak root user
```

The init command also saves the root user password to your `nebari-config.yaml` configuration file under the following path:

`security.keycloak.initial_root_password`

After the initial deployment, it is **highly** recommended that you change the Keycloak `root` user password as soon as you can.

:::note
Once you change the root password you will not be able to [add users from the command line](#add-user-from-the-command-line)

> :::

Reviewer: move this to just below the header for this section.

:::note

> From this point on, the `security.keycloak.initial_root_password` field in `nebari-config.yaml` has no effect. If you redeploy Nebari it will not reset the password back to the old one (or anything else that might be in the field in your YAML file). We strongly recommend you delete this field to prevent later confusion.

Author: moved

:::

1. To change the `root` user password, go to your Nebari instance's admin dashboard - e.g., something like <https://mydomain.com/auth/admin/> — and log in with the password provided.

   ![Nebari admin view - Root Login to Keycloak form](/img/how-tos/keycloak_master_login.png)

2. From there, click on the **Root** dropdown in the top right of the screen, and select **Manage account**.

   ![Keycloak root user page - manage account tab selected](/img/how-tos/keycloak_root_user_manage_account.png)

3. Under **Account Security** click **Signing In**.

   ![Keycloak root user page - account security](/img/how-tos/keycloak_root_user_account_security.png)

4. In the Password section, click the **Update** button. This will guide you through entering your existing root password, and then creating a new password.

   ![Keycloak root user page - account security, update password](/img/how-tos/keycloak_root_user_update_password.png)

## Adding a Nebari user

> You will need to add a Nebari user in order to log in to your Nebari platform. If you have chosen to use GitHub or Auth0 single-sign-on, you must ensure the value you enter in Keycloak under `Username` exactly matches the usernames from GitHub or Auth0, respectively.

Author: Indeed, I added a note regarding this in the authentication section and moved the `adding users` as the first item in the configuration doc.

4. Click **Save**.

If you are using the password authentication provider, you will also need to define a password for the user. Keep in mind that, if using Auth0 or Github the password field is not required. It's best to put the **Temporary** toggle in the **OFF** position. Otherwise the user will be forced to change the password on first login.

![Keycloak add user > credentials tab screenshot - set password](/img/how-tos/keycloak_user_password.png)

### Add user from the command line

To make adding users easier for new Nebari deployments, the following client command will help you with this.

```shell
nebari keycloak -c nebari-config.yaml adduser <username> <password>
```

This will create a new user `<username>` under the `analyst` group, with the initial password provided. Omit the password completely if you are using GitHub or Auth0.

:::note
If you [changed the initial root password for Keycloak](#change-keycloak-root-password) this method will not work.

> :::

Reviewer: move this to just below the section header

## Login to Nebari

Your new user can now log in to Nebari, visit your provided Nebari domain URI which will take you to the login form page and follow the [How to login in to Nebari and start a server](/how-tos/login-keycloak).

![Nebari - Log in to Keycloak page](/img/how-tos/nebari_login_screen.png)

> ## Roles x Groups

Reviewer: ```suggestion
## In-depth look at Roles and Groups
```

| `analyst`   | <ul><li>Conda-Store</li><li>Jupyterhub</li><li>Grafana</li></ul>                        | <ul><li>`conda_store_developer`</li><li>`jupyterhub_developer`</li><li>`grafana_viewer`</li></ul>                             | Blank                   |

To create new groups or modify (or delete) existing groups, log in as `root` and click **Groups** on the left-hand side.

As an example, we create a new group named `conda-store-manager`. This group will have administrator access to the [Conda-Store service].

1. Click **New** in the upper-right hand corner under **Groups**.

![Keycloak groups tab screenshot - user groups view](/img/how-tos/keycloak_groups.png)

- Then, give the new group an appropriate name.

![Keycloak add group form - name field set to conda-store-manager](/img/how-tos/keycloak_new_group1.png)

2. Under **Role Mapping**, add the appropriate **Client Roles** as needed; there should be no need to update the **Realm Roles**.

![Keycloak group conda-store-manager form - role mappings tab focused with expanded client roles dropdown](/img/how-tos/keycloak_new_group2.png)

> In this example, the new group only has one mapped role (`conda_store_admin`); however it's possible to attached multiple **Client Roles** to a single group.

Reviewer: ```suggestion
In this example, the new group only has one mapped role, `conda_store_admin`; however it's possible to attached multiple **Client Roles** to a single group.
```

![Nebari - Log in to Keycloak page](/img/how-tos/nebari_login_screen.png)

## Roles x Groups

Groups represent a collection of users that perform similar actions and therefore require similar permissions. By default, Nebari is deployed with the following groups: ``admin``,
``developer``, ``analyst`` and ``viewer`` (in roughly descending order of power).

| Group       | Access to Nebari Resources                                                                    | Roles | Permissions Description |
| ----------- | ------------------------------------------------------------------------------------------- | ----- | -------------------- |
| ``admin``     | <ul><li>Conda-Store</li><li>Dask</li><li>Jupyterhub</li><li>Grafana</li></ul> |   <ul><li>``conda_store_admin``</li><li>``dask_admin``</li><li>``jupyterhub_admin``</li><li>``grafana_admin``</li></ul> | Blank |
| ``developer`` | <ul><li>Conda-Store</li><li>Dask</li><li>Jupyterhub</li><li>Grafana Developer</li></ul> | <ul><li>``conda_store_developer``</li><li>``dask_developer``</li><li>``jupyterhub_developer``</li><li>``grafana_developer``</li></ul> | Blank |

> | ``analyst``   | <ul><li>Conda-Store</li><li>Jupyterhub</li><li>Grafana</li></ul>  | <ul><li>``conda_store_developer``</li><li>``jupyterhub_developer``</li><li>``grafana_viewer``</li></ul> | Blank |

Author: @costrouc I tried to expand this a litlle bit, to give more info regarding this permission level system, would be happy if you could add some thoughts regarding the missing description for this 'roles'... e.g. what precisely a user with this permission will be able to do, and what not  (for example, the difference between grafana admin and viewer, dask, etc.)

---
id: configuring-keycloak

> title: How to configure Keycloak and add new users

Reviewer 2: ```suggestion
title: Configure Keycloak and add new users
```

description: |
  Configuring Keycloak and adding new users.
---

## Configuring Keycloak

> [Keycloak](https://www.keycloak.org/) is the open-source user management software that's automatically deployed within Nebari. It's used to store the database of all

Reviewer 2: ```suggestion
[Keycloak](https://www.keycloak.org/) is an open source user management software that's automatically deployed within Nebari. It's used to store the database of all
```

either we hyphenate all `open-source` or none of them

users in your Nebari instance, and can provide connectivity to other services such as GitHub/Auth0 single sign-on.

As with many user admin tools, in Keycloak you start with an initial login (`root`) that has the ability to administer and create new users. The `root` user is a Keycloak-specific user. It can only be used to login and manage the Keycloak identity management section of Nebari. It's not a user of the wider Nebari data science platform.

<!-- :::note
Nebari does not create a Keycloak user for you during deployment. If you want to add a new user to Keycloak, follow the steps in the [Adding new users](#adding-a-nebari-user) section below.
::: -->

> 

Reviewer 2: Does this need removing?

4. In the Password section, click the **Update** button. This will guide you through entering your existing root password, and then creating a new password.

   ![Keycloak root user page - account security, update password](/img/how-tos/keycloak_root_user_update_password.png)

:::warning
The `security.keycloak.initial_root_password` field in `nebari-config.yaml` has no effect after changing the `root` password. If you redeploy Nebari it **will not reset** the password back to the old one (or anything else that might be in the field in your YAML file). We strongly recommend you delete this field to prevent later confusion.
:::

## Adding a Nebari user

Nebari does not create a Keycloak user for you during deployment, you will need to add a Nebari user manually in order to log in to your Nebari platform. If you have chosen to use GitHub or Auth0 single-sign-on, you must ensure the value you enter in Keycloak under `Username` exactly matches the usernames from GitHub or Auth0, respectively.

### Add user using Keycloak console

To add a Nebari user from the web console for Keycloak, visit https://{your-nebari-domain}/auth/admin/ and log in using the username `root`, as shown above.

> Keycloak uses [realms](https://www.keycloak.org/docs/latest/server_admin/#configuring-realms) to separate their permission scopes under the Keycloak admin console. The realm for Nebari is `nebari` and all Nebari users will be part of the `nebari` realm, on the other hand, the realm for the Keycloak admin console is `master`.

Reviewer 2: Does it have to be master? I would *highly prefer avoiding non-inclusive terms like master* - could we use something else like: main, captain?
This is a dev-releated broader question so pinging @costrouc

:::note
The root user alone is a member of the `master` realm. For security best practices, we recommend that you keep at minimum any additional users to the `master` realm. See the [Master realm Keycloak documentation](https://www.keycloak.org/docs/latest/server_admin/#the-master-realm) for more information.

> :::

Reviewer 2: same as above flagging for visbility

The `nebari` realm is selected by default, we strongly recommend leaving it as is.

Steps to create a new user:

1. Click **Users** along the left-hand side of the page.

2. Click the **Add user** button and you will see the new user form:

   ![Keycloak add user tab screenshot - new user form](/img/how-tos/keycloak_add_users.png)

> 3. Fill out the three fields outlined above. These are **Username**, **Email**, and **Groups**. (See explanation below). Then click **save**.

Reviewer 2: ```suggestion
3. Fill out the three fields outlined above. These are **Username**, **Email**, and **Groups**. (See explanation in the table below). Then click **save**.
```

      | Username | _unique username_ | _GitHub username_ | _Email to login with_ |

- Once the **Username** field is updated, add a valid email address in the **Email** field.

  :::note
  Although not required, users may not be able to log in to Grafana if this field isn't properly set.
  :::

- Associate the user with one or more of the **Groups**. Out of the box, Nebari is deployed with the following groups: `admin`, `analyst`, and `developer` (see the [Roles x Groups](#roles-x-groups) section below for more details).

4. Click **Save**.

If you are using the password authentication provider, you will also need to define a password for the user. Keep in mind that, if using Auth0 or Github, the password field is not required. It's best to put the **Temporary** toggle in the **OFF** position. Otherwise the user will be forced to change the password on first login.

![Keycloak add user > credentials tab screenshot - set password](/img/how-tos/keycloak_user_password.png)

### Add user from the command line

:::warning
If you [changed the initial root password for Keycloak](#change-keycloak-root-password) this method will not work.

> :::

Reviewer 2: What is the alternative here? It makes no sense that we are telling people to change their password then making a tool that does not work if they follow security best practices 

this is not good user experience and needs fixing cc @iameskild @viniciusdc @costrouc

To make adding users easier for new Nebari deployments, we've created a CLI command to help you.

```shell
nebari keycloak -c nebari-config.yaml adduser <username> <password>
```

This will create a new user `<username>` under the `analyst` group, with the initial password provided. Omit the password completely if you are using GitHub or Auth0.

## Login to Nebari

Your new user can now log in to Nebari, visit your provided Nebari domain URI which will take you to the login form page and follow the [How to login in to Nebari and start a server](/how-tos/login-keycloak).

![Nebari - Log in to Keycloak page](/img/how-tos/nebari_login_screen.png)

## In-depth look at Roles and Groups

Groups represent a collection of users that perform similar actions and therefore require similar permissions. By default, Nebari is deployed with the following groups: `admin`,

> `developer`, `analyst` and `viewer` (in roughly descending order of power).

Reviewer 2: ```suggestion
`developer`, `analyst` and `viewer` (in roughly descending order of permissions and scope).
```

let's avoid non-inclusive terms

| `analyst`   | <ul><li>Conda-Store</li><li>Jupyterhub</li><li>Grafana</li></ul>                        | <ul><li>`conda_store_developer`</li><li>`jupyterhub_developer`</li><li>`grafana_viewer`</li></ul>                             | Blank                   |

To create new groups or modify (or delete) existing groups, log in as `root` and click **Groups** on the left-hand side.

As an example, we create a new group named `conda-store-manager`. This group will have administrator access to the [Conda-Store service].

1. Click **New** in the upper-right hand corner under **Groups**.

![Keycloak groups tab screenshot - user groups view](/img/how-tos/keycloak_groups.png)

- Then, give the new group an appropriate name.

![Keycloak add group form - name field set to conda-store-manager](/img/how-tos/keycloak_new_group1.png)

2. Under **Role Mapping**, add the appropriate **Client Roles** as needed; there should be no need to update the **Realm Roles**.

![Keycloak group conda-store-manager form - role mappings tab focused with expanded client roles dropdown](/img/how-tos/keycloak_new_group2.png)

> In this example, the new group only has one mapped role, `conda_store_admin`; however it's possible to attached multiple **Client Roles** to a single group.

Reviewer 2: ```suggestion
In this example, the new group only has one mapped role, `conda_store_admin`; however, it's possible to attach multiple **Client Roles** to a single group.
```
